app/plugin_install.go

```diff
@@ -142,18 +180,25 @@ const (
 	installPluginLocallyAlways
 )
 
-func (a *App) installPluginLocally(pluginFile io.ReadSeeker, installationStrategy pluginInstallationStrategy) (*model.Manifest, *model.AppError) {
+func (a *App) installPluginLocally(pluginFile, signature io.ReadSeeker, installationStrategy pluginInstallationStrategy) (*model.Manifest, *model.AppError) {
 	pluginsEnvironment := a.GetPluginsEnvironment()
 	if pluginsEnvironment == nil {
 		return nil, model.NewAppError("installPluginLocally", "app.plugin.disabled.app_error", nil, "", http.StatusNotImplemented)
 	}
+	// verify signature
+	if signature != nil {
+		if err := a.VerifyPlugin(pluginFile, signature); err != nil {
+			return nil, err
+		}
+	}
 
 	tmpDir, err := ioutil.TempDir("", "plugintmp")
 	if err != nil {
 		return nil, model.NewAppError("installPluginLocally", "app.plugin.filesystem.app_error", nil, err.Error(), http.StatusInternalServerError)
 	}
 	defer os.RemoveAll(tmpDir)
 
+	pluginFile.Seek(0, 0)
```

Comment:
Doesn't this negate the point of using a stream? Should we just pass around a `[]byte` instead?

Reply:
we actually have both cases, one where we use raw `[]byte` and other places where we have a file to read through. I personally prefer `io.ReadSeeker` interface as its flexible enough to read from various types rather than just `[]byte`. 2/5 that we should continue consuming `io.ReadSeeker` here.

Reply:
Not sure how io.ReadSeeker is more flexible. We can always convert a reader to `[]byte` with `ioutil.ReadAll()`  and go the other way with `bytes.Buffer`

I just find it a bit confusing. It obscures the fact we are buffering anyway.

Reply:
With `io.ReadSeeker` you have the flexibility in the sense that you can easily have an implementation that doesn't load the entire file in memory, but rather buffers it, kinda like `os.File`. However with `[]byte` it forces you to load full content in memory. Granted this approach will not work well with streams :).
Same reason why we have `Reader(path string) (ReadCloseSeeker, *model.AppError)` on `FileBackend`.